### DISCUSSION TITLE
OpenAPI page cannot load if Schema includes lt=inf [BUG]

### DISCUSSION THREAD

**Discussion by jessekv** [fastapi/fastapi#7550] [Questions]
### Describe the bug

Pydantic allows inf, -inf values to be passed to a model. To exclude these, I used the `lt` param:

```python
import math
from pydantic import BaseModel, Field

class MyModel(BaseModel):
    myfloat: float = Field(..., lt=math.inf)
```

The openapi page looks like this:

![image](https://user-images.githubusercontent.com/24944371/79996272-8e089f00-84b8-11ea-8cb9-ba91bc916a9c.png)

And on the server side, a runtime error:
```
...
File "/usr/local/lib/python3.7/site-packages/fastapi/applications.py", line 106, in openapi
return JSONResponse(self.openapi())
File "/usr/local/lib/python3.7/site-packages/starlette/responses.py", line 42, in __init__
self.body = self.render(content)
File "/usr/local/lib/python3.7/site-packages/starlette/responses.py", line 151, in render
separators=(",", ":"),
File "/usr/local/lib/python3.7/json/__init__.py", line 238, in dumps
**kw).encode(obj)
File "/usr/local/lib/python3.7/json/encoder.py", line 199, in encode
chunks = self.iterencode(o, _one_shot=True)
File "/usr/local/lib/python3.7/json/encoder.py", line 257, in iterencode
return _iterencode(o, 0)
ValueError: Out of range float values are not JSON compliant
```

This is a similar issue as #459, except that the invalid value is in the schema itself. The FastAPI.default_response_class is not used to serialize the openapi.json, and this IMO the correct behaviour, but it means the solution proposed in #459 does not apply here. As the model is used in other places outside of fastapi, it does not make sense to remove the `lt` kwarg.

**Comment by jessekv** [comment]
After further investigation, this could be a pydantic issue, as `MyModel.schema()` is:
```json
{
  "properties": {
    "myfloat": {
      "exclusiveMaximum": inf,
      "title": "Myfloat",
      "type": "number"
    }
  },
  "required": [
    "myfloat"
  ],
  "title": "MyModel",
  "type": "object"
}
```
The serialization is failing on the `inf` in the schema. So the error emerges in fastapi, but might best be handled in pydantic. I raised the issue there as well: https://github.com/samuelcolvin/pydantic/issues/1417

**Comment by chris-allnutt** [comment]
@vdwees i'd agree it's a pydantic issue. JSON does allow Infinity, and -Infinity as values, I don't know if those are swagger compliant tho.

This is valid JSON
```
{ "to_infinity_and_beyond": Infinity}
```

**Comment by retnikt** [comment]
±Infinity (nor NaN) are definitely not valid JSON.

**Comment by jessekv** [comment]
@retnikt take a look at https://github.com/samuelcolvin/pydantic/issues/1417

Hope to have this issue fixed soon

**Comment by jessekv** [comment]
https://github.com/samuelcolvin/pydantic/pull/1422 was merged, should this issue be closed now, or should it wait for a pydantic release?

**Comment by tiangolo** [comment]
Thanks for the discussion here everyone! :coffee: 

I think we can close it here as it would be a Pydantic issue more than FastAPI.

And whenever there's a new release, it will be available to be used with FastAPI as well.

So I'm gonna close it now, but feel free to add more comments or new issues.

**Comment by alex-r-bigelow** [comment]
7 months later, and this still seems to be a problem, i.e.:

```python
@router.get('/infinity')
def test_infinity():
    return float('inf')
```
still throws `ValueError: Out of range float values are not JSON compliant`

As far as I can tell, it looks like the pydantic PR was merged and is part of the current version (1.7.3), but I'm not familiar enough with FastAPI's internals to file a sensible bug elsewhere... any hints?

**Comment by danieljfarrell** [comment]
@alex-r-bigelow I have the same problem, did you figure out a work-around?

**Comment by alex-r-bigelow** [comment]
Not yet; for now I'm making deep copies of objects and manually converting any `nan`, `inf`, and `-inf` values to strings

**Comment by danieljfarrell** [comment]
In my case this was because FastAPI was parsing the python object to JSON automatically.

If I set the `response_class` to be `JSONResponse` and manually convert the python object to JSON (for example my using Pydantic) then it works fine. 

I guess FastAPI default parser will not allow illegal JSON values when parsing, who’d a thought it!
